### PR TITLE
cilium-dbg: fix status commands for cluster connectivity health

### LINKED ISSUE #33697: cilium-dbg status: Cluster health reporting unintuitive/incorrect results
### Is there an existing issue for this?

- [X] I have searched the existing issues

### What happened?

When running `cilium-dbg status` within cilium-agent, it outputs cluster health information with a list of nodes, example output:
```
Cluster health:                            10/12 reachable   (2024-06-05T09:38:52Z)
  Name                                     IP                Node          Endpoints
  cluster-1/node-1   172.21.114.19     unreachable   unreachable
```
in this case, it actually did not show one of the unreachable nodes:
```
 cluster-2/node-2               172.22.248.182    unreachable   unreachable
```
even though by default it's supposed to show up to 10 entries and print (...) at the end to indicate partial results.

Thanks @bmcustodio for reporting.

### Cilium Version

main

### Kernel Version

N/A

### Kubernetes Version

N/A

### Regression

No

### Sysdump

_No response_

### Relevant log output

_No response_

### Anything else?

cilium-dbg status is supposed to print up to 10 health lines: https://github.com/cilium/cilium/blob/5f2e61a1cce749619b15aacd957641cf10814a33/cilium-dbg/cmd/status.go#L41

https://github.com/cilium/cilium/blob/5f2e61a1cce749619b15aacd957641cf10814a33/pkg/health/client/client.go#L391C1-L402C3

we can see that actually it does not count the number of lines properly:
- it doesn't take into account if any information was printed or not - this depends on flags succinct/verbose - as in the example output, it only printed a single line even though there was a second node that was also unhealthy. 
- it does not skip counting localhost in this for loop, so localhost is counted twice, first time in https://github.com/cilium/cilium/blob/5f2e61a1cce749619b15aacd957641cf10814a33/pkg/health/client/client.go#L382C1-L385C3
and then also in the loop iterating all nodes

Also, condition for printing "(...)" is incorrect:
https://github.com/cilium/cilium/blob/5f2e61a1cce749619b15aacd957641cf10814a33/pkg/health/client/client.go#L400C1-L402C3
in the case of printing all nodes including healthy nodes, it doesn't take into account for example case, when all nodes are healthy. 
It would result in check `len(sr.Nodes)-healthy = 0 > maxLines` which would be always false.

Proposed design:
Instead of counting and trying to guess if the line was printed or not, let's add boolean returned by `formatNodeStatus` to indicate if it printed a line or not. Once we get maxLines lines printed, we stop printing any remaining lines, but still check if there are remaining nodes, that we didn't iterate over and print "(...)" at the end.
We can also refactor a bit flags of `formatNodeStatus` and get rid of `succinct` flag as it's always set to true.

### Cilium Users Document

- [ ] Are you a user of Cilium? Please add yourself to the [Users doc](https://github.com/cilium/cilium/blob/main/USERS.md)

### Code of Conduct

- [X] I agree to follow this project's Code of Conduct

### PR DESCRIPTION
**Pull request by darox**
This commit fixes serveral issues that were
reported in the status commands of cilium-dbg.

- Fix the issue where not up to 10 unreachable nodes were shown in the status command.
- Fix the issue where `--all-nodes` didn't return all nodes regardless of their status.
- `--all-health` now returns the health of all modules regardless of their status. This can be paired with `--all-nodes` to get the health of all nodes.
- `--verbose` now shows the latency to nodes and endpoints.

Fixes: https://github.com/cilium/cilium/issues/33697

### CODE REVIEW

**Comment by darox** [comment]
Converted back to draft as I work on adding tests for the `FormatHealthStatusResponse` function.

**Review by ldelossa — APPROVED** [review]

File: pkg/health/client/client.go [hunk 1]
```diff
@@ -345,20 +332,53 @@ func formatNodeStatus(w io.Writer, node *models.NodeStatus, printAll, succinct,
 				formatPathStatus(w, "Secondary", addr, "      ", verbose)
 			}
 		}
+		return true
 	}
+
+	if !nodeIsHealthy(node) {
+		ips := []string{getPrimaryAddressIP(node)}
+		for _, addr := range GetHostSecondaryAddresses(node) {
```
Comment (derailed) [inline pkg/health/client/client.go:353]: We should dry this up with something like printNode() so we can reuse the call for unhealthy and all nodes.

File: pkg/health/client/client.go [hunk 2]
```diff
@@ -368,47 +388,45 @@ func FormatHealthStatusResponse(w io.Writer, sr *models.HealthStatusResponse, pr
...
-		fmt.Fprintf(w, "Nodes:\n")
-	}
+
+	fmt.Fprintf(w, "Cluster health:\t%d/%d reachable\t(%s)\n",
+		healthy, len(sr.Nodes), sr.Timestamp)
 
 	if localhost != nil {
-		formatNodeStatus(w, localhost, printAll, succinct, verbose, true)
-		maxLines--
+		printed := formatNodeStatus(w, localhost, allNodes, verbose, true)
```
Comment (derailed) [inline pkg/health/client/client.go]: nit the var declaration is not necessary i.e if formatNodeStatus() {}

File: pkg/health/client/client.go [hunk 3]
```diff
@@ -368,47 +388,45 @@ func FormatHealthStatusResponse(w io.Writer, sr *models.HealthStatusResponse, pr
...
 
 // GetAndFormatHealthStatus fetches the health status from the cilium-health
 // daemon via the default channel and formats its output as a string to the
 // writer.
 //
-// 'succinct', 'verbose' and 'maxLines' are handled the same as in
+// 'verbose' and 'maxLines' are handled the same as in
 // FormatHealthStatusResponse().
-func GetAndFormatHealthStatus(w io.Writer, succinct, verbose bool, maxLines int) {
+func GetAndFormatHealthStatus(w io.Writer, allNodes bool, verbose bool, maxLines int) {
```
Comment (derailed) [inline pkg/health/client/client.go:452]: Prefer allNodes, verbose bool

File: pkg/health/client/client.go [hunk 4]
```diff
@@ -345,20 +332,53 @@ func formatNodeStatus(w io.Writer, node *models.NodeStatus, printAll, succinct,
...
 // writer.
 //
-// 'printAll', if true, causes all nodes to be printed regardless of status
-// 'succinct', if true, causes node health to be output as one line per node
-// 'verbose', if true, overrides 'succinct' and prints all information
+// 'allNodes', if true, causes all nodes to be printed regardless of status
+// 'verbose', if true, prints all information
 // 'maxLines', if nonzero, determines the maximum number of lines to print
-func FormatHealthStatusResponse(w io.Writer, sr *models.HealthStatusResponse, printAll, succinct, verbose bool, maxLines int) {
+func FormatHealthStatusResponse(w io.Writer, sr *models.HealthStatusResponse, allNodes bool, verbose bool, maxLines int) {
```
Comment (derailed) [inline pkg/health/client/client.go:400]: Likely in a separate pr but might be a good idea to just pass in a printOptions struct so we can extend in the future without the need to alter the function ie FormatHealthStatusResponse(io.Writer, HealthStatusResp, PrintOpts)

**Review by derailed — CHANGES_REQUESTED** [review]
@darox Nice work!

File: pkg/health/client/client.go (hunk 1 above)
Comment (darox) [reply pkg/health/client/client.go:353]: Good idea. Implemented.

File: pkg/health/client/client.go (hunk 2 above)
Comment (darox) [reply pkg/health/client/client.go]: Very valuable recommendation that makes the code even easier. Implemented.

File: pkg/health/client/client.go (hunk 3 above)
Comment (darox) [reply pkg/health/client/client.go:452]: Yes, I didn't know that Go has this kind of grouping for data types. Thank you.

File: pkg/health/client/client.go (hunk 4 above)
Comment (darox) [reply pkg/health/client/client.go:400]: Good idea. I will start looking into this after this PR gets merged.

**Comment by darox** [comment]
@derailed I pushed the changes you've requested. I will rebase once you're OK with the changes.

**Review by derailed — APPROVED** [review]
@darox Thank you for the updates!

**Comment by marseel** [comment]
@darox seems like some tests are failing, would you mind rebasing PR, hopefully it fixes it :smiley: ?

**Comment by darox** [comment]
> @darox seems like some tests are failing, would you mind rebasing PR, hopefully it fixes it 😃 ?

I need a bit more time, because there's now a conflict:


```
+++ b/pkg/health/client/client.go
@@@ -325,29 -315,8 +325,34 @@@ func formatNodeStatus(w io.Writer, nod
        if localhost {
                localStr = " (localhost)"
        }
++<<<<<<< HEAD
 +      if succinct {
 +              if printAll || !nodeIsHealthy(node) {
 +                      ips := []string{getPrimaryAddressIP(node)}
 +                      for _, addr := range GetHostSecondaryAddresses(node) {
 +                              if addr == nil {
 +                                      continue
 +                              }
 +                              ips = append(ips, addr.IP)
 +                      }
 +                      hostStatuses := SummarizePathConnectivityStatusType(GetAllHostAddresses(node))
 +                      endpointStatuses := SummarizePathConnectivityStatusType(GetAllEndpointAddresses(node))
 +
 +                      fmt.Fprintf(w, "  %s%s\t%s\t%d/%d", node.Name, localStr, strings.Join(ips, ","), hostStatuses[ConnStatusReachable], len(GetAllHostAddresses(node)))
 +                      if hostStatuses[ConnStatusUnknown] > 0 {
 +                              fmt.Fprintf(w, " (%d unknown)", hostStatuses[ConnStatusUnknown])
 +                      }
 +                      fmt.Fprintf(w, "\t%d/%d", endpointStatuses[ConnStatusReachable], len(GetAllEndpointAddresses(node)))
 +                      if endpointStatuses[ConnStatusUnknown] > 0 {
 +                              fmt.Fprintf(w, " (%d unknown)", endpointStatuses[ConnStatusUnknown])
 +                      }
 +                      fmt.Fprintf(w, "\n")
 +              }
 +      } else {
++=======
+
```

**Comment by darox** [comment]
I rebased and fixed the conflicts.

**Comment by marseel** [comment]
I've re-requested  @ldelossa review due to the conflicts you mentioned.

**Comment by darox** [comment]
I did a bit more testing and noticed the following:

- With the current solution (main branch), we're not outputting the secondary IP health status:

```
Cluster health:	3/4 reachable	(2023-04-01T12:00:00Z)
Name	IP	Node	Endpoints
  node0:
    Host connectivity to 192.168.1.0:
      ICMP to stack:	OK, RTT=10ms
      HTTP to agent:	OK, RTT=10ms
    Endpoint connectivity to 192.168.1.0:
      ICMP to stack:	OK, RTT=10ms
      HTTP to agent:	OK, RTT=10ms
  node1:
    Host connectivity to 192.168.1.1:
      ICMP to stack:	OK, RTT=10ms
      HTTP to agent:	OK, RTT=10ms
    Endpoint connectivity to 192.168.1.1:
      ICMP to stack:	OK, RTT=10ms
      HTTP to agent:	OK, RTT=10ms
  node2:
    Host connectivity to 192.168.1.2:
      ICMP to stack:	OK, RTT=10ms
      HTTP to agent:	OK, RTT=10ms
    Endpoint connectivity to 192.168.1.2:
      ICMP to stack:	OK, RTT=10ms
      HTTP to agent:	OK, RTT=10ms
  node3:
    Host connectivity to 192.168.1.3:
      ICMP to stack:	failed
      HTTP to agent:	failed
    Endpoint connectivity to 192.168.1.3:
      ICMP to stack:	failed
      HTTP to agent:	failed
```

But with my PR it will be:
```
Name              IP              Node   Endpoints
  clustermesh2/clustermesh2-worker (localhost):
    Host connectivity to 172.19.0.5:
      ICMP to stack:   OK, RTT=1.52943ms
      HTTP to agent:   OK, RTT=64.135µs
      Secondary connectivity to fc00:c111::5:
        ICMP to stack:   OK, RTT=1.66545ms
        HTTP to agent:   OK, RTT=66.384µs
    Endpoint connectivity to 10.2.1.114:
      ICMP to stack:   OK, RTT=1.648948ms
      HTTP to agent:   OK, RTT=166.983µs
      Secondary connectivity to fd00:10:2:1::2e07:
        ICMP to stack:   OK, RTT=1.600523ms
        HTTP to agent:   OK, RTT=139.687µs
  clustermesh1/clustermesh1-control-plane:
    Host connectivity to 172.19.0.2:
      ICMP to stack:   OK, RTT=1.630403ms
      HTTP to agent:   OK, RTT=318.88µs
      Secondary connectivity to fc00:c111::2:
        ICMP to stack:   OK, RTT=1.614359ms
        HTTP to agent:   OK, RTT=233.326µs
    Endpoint connectivity to 10.1.0.55:
      ICMP to stack:   OK, RTT=2.948719ms
      HTTP to agent:   OK, RTT=621.549µs
      Secondary connectivity to fd00:10:1::7394:
        ICMP to stack:   OK, RTT=1.62011ms
        HTTP to agent:   OK, RTT=349.884µs
  clustermesh1/clustermesh1-worker:
    Host connectivity to 172.19.0.3:
      ICMP to stack:   OK, RTT=2.978723ms
      HTTP to agent:   OK, RTT=154.272µs
      Secondary connectivity to fc00:c111::3:
        ICMP to stack:   OK, RTT=1.671075ms
        HTTP to agent:   OK, RTT=113.725µs
    Endpoint connectivity to 10.1.1.211:
      ICMP to stack:   OK, RTT=1.595981ms
      HTTP to agent:   OK, RTT=262.663µs
      Secondary connectivity to fd00:10:1:1::dcc9:
        ICMP to stack:   OK, RTT=1.617026ms
        HTTP to agent:   OK, RTT=352.343µs
  clustermesh2/clustermesh2-control-plane:
    Host connectivity to 172.19.0.4:
      ICMP to stack:   OK, RTT=2.956429ms
      HTTP to agent:   OK, RTT=135.436µs
      Secondary connectivity to fc00:c111::4:
        ICMP to stack:   OK, RTT=1.68337ms
        HTTP to agent:   OK, RTT=112.933µs
    Endpoint connectivity to 10.2.0.132:
      ICMP to stack:   OK, RTT=1.631987ms
      HTTP to agent:   OK, RTT=306.67µs
      Secondary connectivity to fd00:10:2::85f:
        ICMP to stack:   OK, RTT=1.578645ms
        HTTP to agent:   OK, RTT=271.414µs
```

Does that make sense?

**Comment by darox** [comment]
I don't see how this PR would affect the failing CI task.

**Comment by marseel** [comment]
I've rerun test, let's see if it was just a flake (probably)

**Comment by darox** [comment]
> I've rerun test, let's see if it was just a flake (probably)

passed

**Comment by darox** [comment]
@ldelossa can we get it merged?

**Comment by marseel** [comment]
@ldelossa would you mind taking a look or should we reassign?

**Review by ldelossa — APPROVED** [review]
LGTM

**Comment by darox** [comment]
What are the steps I have to take, to get this backported to v1.16?

**Comment by giorio94** [comment]
> What are the steps I have to take, to get this backported to v1.16?

Hey @darox :wave: Are you still willing to take care of this author backport? 

If so, the easiest approach is probably to:
1. Check out the v1.16 branch
2. Execute `./contrib/backporting/cherry-pick bf7bc43bd53ca77934513ffd887a59974deca24c`
3. Fix the conflicts
4. Run `git am --continue`
5. Amend the commit to add a description of the addressed conflicts (e.g., https://github.com/cilium/cilium/pull/37281/commits/ce332556e477d06cd3f565fed3c426192a6c1bd5)
6. Open a PR against the v1.16 branch.
7. Once merged, change the label of this one to `backport-done/v1.16`

Feel free to reach out to me if you have questions.

**Comment by darox** [comment]
@giorio94 If I will find time, I will take a look.

**Comment by darox** [comment]
I had some time to check the code and I think we first need to apply https://github.com/cilium/cilium/commit/fee61079411bd1c32b678ddf00601b24bfb7e90b before I can apply my commit. My whole commit is based on the former commit. Now, the question is, if we can apply https://github.com/cilium/cilium/commit/fee61079411bd1c32b678ddf00601b24bfb7e90b because it deprecates some metrics. This commit was first applied on 1.17. I pushed it to https://github.com/cilium/cilium/pull/37463 which fixes it. But someone will have to decide, whether that's fine.

**Comment by darox** [comment]
> > Now, the question is, if we can apply https://github.com/cilium/cilium/commit/fee61079411bd1c32b678ddf00601b24bfb7e90b because it deprecates some metrics.
> 
> 
> 
> That commit introduces a new feature, which is typically not eligible for being backported, and IMO it would come with a significant risk for stable releases. If that commit is a strict requirement to backport this PR, my feeling is that it would be better skipping this backport, considering that it fixes a CLI command that most likely always suffered from this bug.

Yes, I would agree with that.

**Comment by giorio94** [comment]
Ack, labels dropped.

**Pull request merged**